code/controllers/subsystem/gravity.dm

```diff
@@ -0,0 +1,92 @@
+var/datum/subsystem/gravity/SSgravity
+var/global/legacy_gravity = FALSE
+
+/datum/subsystem/gravity
+	name = "Gravity"
+	priority = 75
+	wait = 2
+	init_order = -100
+	flags = SS_KEEP_TIMING | SS_BACKGROUND
+
+	var/list/currentrun = list()
+	var/list/currentrun_manual = list()
+	var/list/areas = list()
+	var/recalculation_cost = 0
+	var/purge_interval = -1
+	var/purge_tick = 0
+	var/purging = FALSE
+	var/list/purging_atoms = list()
+
+/datum/subsystem/gravity/New()
+	NEW_SS_GLOBAL(SSgravity)
+
+/datum/subsystem/gravity/Initialize()
+	for(var/area/A in world)
+		areas += A
+	. = ..()
+
+/datum/subsystem/gravity/proc/recalculate_atoms()
+	currentrun = list()
+	currentrun_manual = list()
+	var/tempcost = world.timeofday
+	for(var/area/A in areas)
+		if(!A.has_gravity && !A.gravity_generator && !A.gravity_overriding)
+			continue
+		if(!A.gravity_direction)
+			continue
+		for(var/atom/movable/AM in A.contents_affected_by_gravity)
+			if(AM.force_gravity_processing)
+				continue
+			currentrun += AM
```

Comment:
Instead of doing all these checks in this proc, do it in the `fire()` loop and skip things there